UNR-2318 stop spatial service which runs in different project

v1ctor opened the pull request:
#### Description
In this diff when user start a local deployment I check that it's running in the correct directory. If it's running from a wrong directory I show dialog and stop the service which is already running.  

#### Release note
REQUIRED: Add a release note to the `##Unreleased` section of CHANGELOG.md. You can find guidance for writing useful release notes [here](../SpatialGDK/Extras/internal-documentation/how-to-write-good-release-notes.md). Documentation changes are exempt from this requirement.

#### Tests
1. Start a spatialos service from another project. 
2. Go to the target project and try to start a local deployment.
3. See the dialogue to stop an already running service. 
4. Click yes
5. Service which is running in a wrong project is stopped. Local deployment and spatial service started successfully. 

#### Documentation
Do we need a documentation here?
We should check a dialogue window text.

#### Primary reviewers
@joshuahuburn

joshuahuburn commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -273,6 +273,17 @@ bool FLocalDeploymentManager::LocalDeploymentPreRunChecks()
 		}
 	}
 
+	if (!IsServiceRunningAndInCorrectDirectory()) {
```
```suggestion
	if (!IsServiceRunningAndInCorrectDirectory()) 
	{
```

joshuahuburn commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -273,6 +273,17 @@ bool FLocalDeploymentManager::LocalDeploymentPreRunChecks()
 	}
 
+	if (!IsServiceRunningAndInCorrectDirectory()) {
+		if (FMessageDialog::Open(EAppMsgType::YesNo, LOCTEXT("StopSpatailServiceFromDifferentProject", "Another SpatialOS service is already running in another project. Would you like to stop it?")) == EAppReturnType::Yes)
```
@jessicafalk could we confirm this is fine with Xcode?

jessicafalk replied:
that should be fine ;) no exclamation mark that can cause confusion

m-samiec commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -269,6 +269,18 @@ bool FLocalDeploymentManager::LocalDeploymentPreRunChecks()
 		}
 	}
 
+	if (!IsServiceRunningAndInCorrectDirectory()) 
```
Will this return false if the server isn't running at all?

v1ctor replied:
@m-samiec It's a good catch, let me double check this.

v1ctor commented:
I've refactored code a little bit and make it to check that service is actually running before checking for the wrong directory.
I've tested this solution on following test cases:
1. Service is running in a wrong directory
  1. We write an error message to the output log.
  2. If you press start local deployment we show a dialog to stop a wrong service
2. Service isn't running at all
  1. Service and deployment start normally 
3. Service running in the same project
  1. We do nothing

I've extracted call to `spot alpha service` to only execute it once. otherwise I'd need to call it to check if it's running and the second time to get a project name.

m-samiec commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -623,20 +635,16 @@ bool FLocalDeploymentManager::GetLocalDeploymentStatus()
-		return false;
-	}
-
+FSpotServiceInfoResult FLocalDeploymentManager::GetServiceInfo() {
```
```suggestion
FSpotServiceInfoResult FLocalDeploymentManager::GetServiceInfo()
{
```

m-samiec commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -648,60 +656,74 @@ bool FLocalDeploymentManager::IsServiceRunningAndInCorrectDirectory()
+		FPaths::RemoveDuplicateSlashes(Result.ProjectFileName);
+		Result.bServiceIsRunning = true;
+	}
+	else {
```
```suggestion
	else
	{
```

m-samiec commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -648,60 +656,74 @@ bool FLocalDeploymentManager::IsServiceRunningAndInCorrectDirectory()
-		{
-			UE_LOG(LogSpatialDeploymentManager, Error,
-				TEXT("Spatial service running in a different project! Please run 'spatial service stop' if you wish to launch deployments in the current project. Service at: %s"), *SpatialServiceProjectPath);
+bool FLocalDeploymentManager::IsServiceRunningAndInCorrectDirectory(const FString& ServiceProjectName) {
```
```suggestion
bool FLocalDeploymentManager::IsServiceRunningAndInCorrectDirectory(const FString& ServiceProjectName)
{
```

m-samiec commented on SpatialGDK/Source/SpatialGDKServices/Public/LocalDeploymentManager.h:
```diff
@@ -13,6 +13,11 @@ DECLARE_LOG_CATEGORY_EXTERN(LogSpatialDeploymentManager, Log, All);
 
 class FJsonObject;
 
+struct FSpotServiceInfoResult {
```
```suggestion
struct FSpotServiceInfoResult
{
```

joshuahuburn commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -287,6 +286,19 @@ bool FLocalDeploymentManager::LocalDeploymentPreRunChecks()
 		}
 	}
 
+	FSpotServiceInfoResult ServiceInfo = GetServiceInfo();
```
Just realised. Is this adding any significant (greater than 0.5s) time delay to starting a local deployment? 
Since this calling an executable now it might slow down day to day development. We have an NFR for deployment launching taking no more than 5s in total from clicking start to the deployment being ready (we are already at 4.5s)
Could you please check how much time this is adding :D

m-samiec commented on SpatialGDK/Source/SpatialGDKServices/Private/LocalDeploymentManager.cpp:
```diff
@@ -287,6 +287,18 @@ bool FLocalDeploymentManager::LocalDeploymentPreRunChecks()
 
+	if (!bSpatialServiceInProjectDirectory)
+	{
+		if (FMessageDialog::Open(EAppMsgType::YesNo, LOCTEXT("StopSpatailServiceFromDifferentProject", "Another SpatialOS service is already running in another project. Would you like to stop it?")) == EAppReturnType::Yes)
```
```suggestion
		if (FMessageDialog::Open(EAppMsgType::YesNo, LOCTEXT("StopSpatialServiceFromDifferentProject", "Another SpatialOS service is already running in another project. Would you like to stop it?")) == EAppReturnType::Yes)
```
